### PR TITLE
New property to always AllowUnauthACFUpload

### PR DESCRIPTION
**Pull request by joseph-reynolds**
This adds a new BMC security setting as a Redfish OEM property: URI /redfish/v1/AccountService/Accounts/service property Oem.IBM.ACF.AllowUnauthACFUpload can be set by the admin.  When true, any network agent is allowed to upload an ACF file to the BMC.  The default value is false.  This capability is included in the existing URI /redfish/v1 property Oem.IBM.ACFWindowActive.  For example, AllowUnauthACFUpload=true implies ACFWindowActive=true.

Tested: via Redfish scenarios
1. Ensure AllowUnauthACFUpload=false by default.
2. Ensure a non admin user is not allowed to GET or PATCH AllowUnauthACFUpload.
3. Ensure the admin can PATCH AllowUnauthACFUpload true and false, and GET the correct values, and these values survive BMC reboot.
4. With (OpPanel fn 74 is not active) and AllowUnauthACFUpload=false, ensure ACFWindowActive=false.  And unauth ACF upload is not allowed.
5. With (OpPanel fn 74 is not active) and AllowUnauthACFUpload=true, ensure ACFWindowActive=true.  And unauth ACF upload is allowed.

### CODE REVIEW

**Comment by joseph-reynolds** [comment]
This fixes the BMCWeb / Redfish portion of SW553048.  The Web GUI will go on top of this.
I will run the Redfish validator for the new property.

File: redfish-core/lib/account_service.hpp [hunk 1]
```diff
@@ -994,6 +996,74 @@ inline void
...
 }
 
+// This file holds the value of the AllowUnauthACFUpload property.  An
+// Administrator user can GET and PATCH this property.  Contrast with the
+// readonly Oem.IBM.ACFWindowActive property (in /redfish/v1) which is true
+// when either AllowUnauthACFUpload or the "OpPanel function 74 time window"
+// is active.  Low-level design:
+//  - When this file exists, AllowUnauthACFUpload=True.
+//  - Otherwise, AllowUnauthACFUpload=False (the default value).
+const std::filesystem::path allowUnauthACFUploadFilename{"/var/lib/AllowUnauthACFUpload"};
```
Comment (human3rr) [inline redfish-core/lib/account_service.hpp:1006]: Is it sound for us to be creating this file from bmcweb for the management of this setting? It just seems unconventional since we're normally using dbus getters and setters for properties normally.

Comment (joseph-reynolds) [reply redfish-core/lib/account_service.hpp:1006]: Correct: this does not follow the pattern of having a D-Bus service.  I looked for a service this setting would go into but did not find one.  And I don't want to create a new one just for this.

Using a file as backing store for a BMC setting is a sound design (AFAIK) and Adriana suggested using /var/lib.  The unconventional element is doing this directly from BMCWeb and not a service.

**Comment by joseph-reynolds** [comment]
Sorry, the changes in account_service.hpp are hard to follow.  The first commit has a refactor and the changes.  The second commit is just a code reformat (and the diff lines up poorly).

**Comment by gtmills** [comment]
@joseph-reynolds have a defect here?

File: redfish-core/lib/service_root.hpp [hunk 2]
```diff
@@ -20,12 +20,22 @@
 #include <registries/privilege_registry.hpp>
 #include <utils/systemd_utils.hpp>
 
+#include <filesystem>
```
Comment (human3rr) [inline redfish-core/lib/service_root.hpp:23]: Is this include required here?

File: redfish-core/lib/account_service.hpp [hunk 3]
```diff
@@ -994,6 +996,74 @@ inline void
...
+
+inline bool SetPropertyAllowUnauthACFUpload(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp, bool allow)
+{
+    if (allow)
+    {
+        // Create the file to indicate the property is true
+        std::ofstream file;
+        file.open(allowUnauthACFUploadFilename);
```
Comment (human3rr) [inline redfish-core/lib/account_service.hpp:1016]: I don't see anywhere that we close the file

Comment (joseph-reynolds) [reply redfish-core/lib/account_service.hpp:1016]: The ofstream object is destructed when it does out of context.  Having the destructor close the file is the recommended usage pattern.  (Should I use some other code to create the file?)  Overall, this code creates/opens the file, change its mode to 600, and closes the file.  (I copy/pasted it from elsewhere in BMCWeb.)

File: redfish-core/lib/service_root.hpp (hunk 2 above)
Comment (joseph-reynolds) [reply redfish-core/lib/service_root.hpp:23]: I think ofstream is provided by filesystem.

**Comment by joseph-reynolds** [comment]
> @joseph-reynolds have a defect here?

SW553048.  This fixes the BMCWeb / Redfish portion of SW553048.  The Web GUI will go on top of this.

**Review by human3rr — APPROVED** [review]

File: redfish-core/lib/account_service.hpp (hunk 3 above)
Comment (human3rr) [reply redfish-core/lib/account_service.hpp:1016]: No, it's fine. I wasn't sure if it the file was closed on fd destructure.

**Comment by joseph-reynolds** [comment]
The Redfish validator passed https://sys-openbmc-dev-jenkins.swg-devops.com/job/OpenBMC-Dev/job/test/job/continuous-integration/job/ebmc/job/Redfish-Service-Validator/985/

File: redfish-core/lib/service_root.hpp (hunk 2 above)
Comment (gtmills) [reply redfish-core/lib/service_root.hpp:23]: but where is ofstream in this file?

File: redfish-core/lib/account_service.hpp (hunk 1 above)
Comment (gtmills) [reply redfish-core/lib/account_service.hpp]: I really think we should follow the architecture of bmcweb and OpenBMC which is Redfish <-> D-Bus

File: redfish-core/lib/service_root.hpp [hunk 4]
```diff
@@ -20,12 +20,22 @@
...
+
 namespace redfish
 {
 
 inline void
     handleACFWindowActive(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
 {
+    bool allow;
+    getPropertyAllowUnauthACFUpload(asyncResp, allow); // ignore return code
```
Comment (gtmills) [inline redfish-core/lib/service_root.hpp:32]: Why would you not just return turn/false for if it was allowed?

Comment (gtmills) [reply redfish-core/lib/service_root.hpp:32]: E.g. https://en.cppreference.com/w/cpp/filesystem/is_regular_file

Return value

true if the file indicated by p or if the type indicated by s refers to a regular file, false otherwise. The non-throwing overload returns false if an error occurs. 


If you want the ec out, pass it back..

File: redfish-core/lib/account_service.hpp [hunk 5]
```diff
@@ -994,6 +996,82 @@ inline void
...
+                             << ") ec=" << ec;
+            asyncResp->res = {};
+            messages::internalError(asyncResp->res);
+            return false;
+        }
+    }
+    return true;
+}
+
+inline bool getPropertyAllowUnauthACFUpload(
```
Comment (gtmills) [inline redfish-core/lib/account_service.hpp:1058]: Why can't this be 
```
inline bool allowUnauthACFUpload(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
{
     std::error_code ec;
// Use noexcept but ignore the ec
return std::filesystem::is_regular_file(allowUnauthACFUploadFilename, ec);
} 
```

or 
```
inline bool allowUnauthACFUpload(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
{
     std::error_code ec;
// Use noexcept 
allow = std::filesystem::is_regular_file(allowUnauthACFUploadFilename, ec);

if (ec)
{
         BMCWEB_LOG_ERROR ...
         messages::internalError(asyncResp->res);
         return false;
}
return allow;

} 
```

Why this crazy logic ?

File: redfish-core/lib/account_service.hpp [hunk 6]
```diff
@@ -994,6 +996,82 @@ inline void
...
 
+// This file holds the value of the AllowUnauthACFUpload property.  An
+// Administrator user can GET and PATCH this property.  Contrast with the
+// readonly Oem.IBM.ACFWindowActive property (in /redfish/v1) which is true
+// when either AllowUnauthACFUpload or the "OpPanel function 74 time window"
+// is active.  Low-level design:
+//  - When this file exists, AllowUnauthACFUpload=True.
+//  - Otherwise, AllowUnauthACFUpload=False (the default value).
+const std::filesystem::path allowUnauthACFUploadFilename{
+    "/var/lib/AllowUnauthACFUpload"};
```
Comment (gtmills) [inline redfish-core/lib/account_service.hpp:1007]: This gets blasted on a factory reset but would serve a code update/reboot?

**Comment by joseph-reynolds** [comment]
Note: I paused this to rework the new property into a D-Bus interface.  I will push new changes when that is ready.

**Review by gtmills — CHANGES_REQUESTED** [review]
Need changes to use D-Bus

File: redfish-core/lib/account_service.hpp (hunk 6 above)
Comment (joseph-reynolds) [reply redfish-core/lib/account_service.hpp]: Correct.  That is the intention.
Note this code is re-written in the next commit to use a new D-Bus property.  That also survives BMC reboot but is reset to default via BMC factory reset.

File: redfish-core/lib/service_root.hpp (hunk 4 above)
Comment (joseph-reynolds) [reply redfish-core/lib/service_root.hpp]: I have rewritten to simplify.

File: redfish-core/lib/account_service.hpp [hunk 7]
```diff
@@ -1596,91 +1644,126 @@ inline void triggerUnauthenticatedACFUpload(
...
     if (ibm)
     {
-        std::optional<nlohmann::json> acf;
         if (!redfish::json_util::readJson(*ibm, asyncResp->res, "ACF", acf))
         {
             BMCWEB_LOG_ERROR << "Illegal Property ";
             messages::propertyMissing(asyncResp->res, "ACF");
             return;
         }
+    }
```
Comment (joseph-reynolds) [inline redfish-core/lib/account_service.hpp:1657]: I flattened this section of code, but it is mostly the same.

File: redfish-core/lib/account_service.hpp [hunk 8]
```diff
@@ -2318,209 +2401,220 @@ inline void requestAccountServiceRoutes(App& app)
...
-            bool isUserItself = false;
-
-            if (!json_util::readJson(
-                    req, asyncResp->res, "UserName", newUserName, "Password",
-                    password, "RoleId", roleId, "Enabled", enabled, "Locked",
-                    locked, "Oem", oem, "AccountTypes", accountType))
-            {
-                return;
-            }
+        .methods(boost::beast::http::verb::patch)(
```
Comment (joseph-reynolds) [inline redfish-core/lib/account_service.hpp:2404]: I don't know why this has changed.  Maybe I need to rebase?

**Comment by joseph-reynolds** [comment]
The code looks okay after the rebase and 3-way merge.  Please review.

File: redfish-core/lib/account_service.hpp [hunk 9]
```diff
@@ -1680,91 +1726,126 @@ inline void triggerUnauthenticatedACFUpload(
         return;
     }
 
+    std::optional<nlohmann::json> acf;
```
Comment (gtmills) [inline redfish-core/lib/account_service.hpp:1729]: Why change this?

File: redfish-core/lib/account_service.hpp [hunk 10]
```diff
@@ -1680,91 +1726,126 @@ inline void triggerUnauthenticatedACFUpload(
...
     if (ibm)
     {
-        std::optional<nlohmann::json> acf;
         if (!redfish::json_util::readJson(*ibm, asyncResp->res, "ACF", acf))
         {
             BMCWEB_LOG_ERROR << "Illegal Property ";
             messages::propertyMissing(asyncResp->res, "ACF");
             return;
         }
+    }
```
Comment (gtmills) [inline redfish-core/lib/account_service.hpp:1738]: This doesn't look right, see: https://github.com/openbmc/bmcweb/blob/613dabea03cd2f5cb999be37aaf539280c63ea7a/redfish-core/lib/managers.hpp#L2231

See how it is 
```
        if (oem)   <- layer 1
        {
               std::optional<nlohmann::json> openbmc; <- now let's peel layer 2 
              if (!redfish::json_util::readJson(*oem, asyncResp->res, "OpenBmc",
                                              openbmc))
            {
                BMCWEB_LOG_ERROR
                    << "Illegal Property "
                    << oem->dump(2, ' ', true,
                                 nlohmann::json::error_handler_t::replace);
                return;
            }
            if (openbmc)
            {
                std::optional<nlohmann::json> fan;
                if (!redfish::json_util::readJson(*openbmc, asyncResp->res,
                                                  "Fan", fan))                                              
                                              
```

Let's follow that pattern

File: redfish-core/lib/account_service.hpp [hunk 11]
```diff
@@ -1680,91 +1726,126 @@ inline void triggerUnauthenticatedACFUpload(
...
 
-        if (acf)
+    std::vector<uint8_t> decodedAcf;
+    if (acf)
```
Comment (gtmills) [inline redfish-core/lib/account_service.hpp:1741]: This should all be within the if(ibm) which should inside of a if(oem)

File: redfish-core/lib/account_service.hpp [hunk 12]
```diff
@@ -1176,6 +1199,26 @@ inline void getAcfProperties(
...
+    crow::connections::systemBus->async_method_call(
+        [asyncResp](const boost::system::error_code ec, 
+                    const std::variant<bool>& retval) {
+            if (ec)
+            {
+                BMCWEB_LOG_ERROR << "D-Bus responses error: " << ec;
+                messages::internalError(asyncResp->res);
+                return;
+            }
+            const bool *allowed = std::get_if<bool>(&retval);
```
Comment (gtmills) [inline redfish-core/lib/account_service.hpp:1212]: check if nullptr

**Comment by joseph-reynolds** [comment]
This ill-fated PR was moved to 1030 PR https://github.com/ibm-openbmc/bmcweb/pull/466 where it merged, and then it was backported to 1020 via https://github.com/ibm-openbmc/bmcweb/commit/921ba85e0bc68779573db121790bde574332d395.  Closing this.
